kotori/daq/storage/cratedb.py

```diff
@@ -0,0 +1,270 @@
+
+        self.host_uri = "{host}:{port}".format(**self.__dict__)
+
+        # TODO: Bring back pool size configuration.
+        # log.info('Storage target is {uri}, pool size is {pool_size}', uri=self.host_uri, pool_size=self.pool_size)
+        log.info("Storage target is {uri}", uri=self.host_uri)
+        self.db_client = client.connect(
+            self.host_uri, username=self.username, password=self.password, pool_size=20,
+        )
+
+    def get_tablename(self, meta):
+        """
+        Get table name for SensorWAN channel.
+        """
+        return f"{meta.database}.{meta.measurement}"
+
+    def create_table(self, tablename):
+        """
+        Create database table for SensorWAN channel.
+        """
+        log.info(f"Creating table: {tablename}")
+        sql_ddl = f"""
+CREATE TABLE IF NOT EXISTS {tablename} (
+    time TIMESTAMP WITH TIME ZONE DEFAULT NOW() NOT NULL, 
+    tags OBJECT(DYNAMIC), 
+    fields OBJECT(DYNAMIC)
+);      
+        """
+        cursor = self.db_client.cursor()
+        cursor.execute(sql_ddl)
+        cursor.close()
+
+    def query(self, expression: str, tdata: Munch = None):
+        """
+        Query CrateDB and respond with results in suitable shape.
+
+        Make sure to synchronize data by using `REFRESH TABLE ...` before running
+        the actual `SELECT` statement. This is applicable in test case scenarios.
+
+        Response format::
+
+            [
+              {
+                "time": ...,
+                "tags": {"city": "berlin", "location": "balcony"},
+                "fields": {"temperature": 42.42, "humidity": 84.84},
+              },
+              ...
+            ]
+
+        TODO: Unify with `kotori.test.util:CrateDBWrapper.query`.
+        """
+
+        log.info(f"Database query: {expression}")
+
+        tdata = tdata or {}
+
+        # Before reading data from CrateDB, synchronize it.
+        # Currently, it is mostly needed to satisfy synchronization constraints when running the test suite.
+        # However, users also may expect to see data "immediately". On the other hand, in order to satisfy
+        # different needs, this should be made configurable per realm, channel and/or request.
+        # TODO: Maybe just _optionally_ synchronize with the database when reading data.
+        if tdata:
+            refresh_sql = f"REFRESH TABLE {self.get_tablename(tdata)}"
+            self.execute(refresh_sql)
+
+        def dict_from_row(columns, row):
+            """
+            https://stackoverflow.com/questions/3300464/how-can-i-get-dict-from-sqlite-query
+            https://stackoverflow.com/questions/4147707/python-mysqldb-sqlite-result-as-dictionary
+            """
+            return dict(zip(columns, row))
+
+        def record_from_dict(item):
+            record = OrderedDict()
+            record.update({"time": item["time"]})
+            record.update(item["tags"])
+            record.update(item["fields"])
+            return record
```

Comment:
I was looking for a function to merge those two objects, `tags` and `fields`, into a single record, but have not been able to discover it. Only now, I luckily discovered the [`concat(object, object)` function](https://crate.io/docs/crate/reference/en/5.3/general/builtins/scalar-functions.html#concat-object-object) in @proddata's tutorial [^1], and that it can also operate on objects, effectively merging those.

I think the two reasons why I have not been able to discover this function were:

a) That both sets of functions operating on container data types [^2][^3] have been on the "Scalar functions" page, and I did not expect to find them there.
b) That the search term "merge" did not occur in the corresponding documentation section of the `concat(object, object)` function, contrary to the documentation of the `array_unique()` function.

   > The `concat(object, object)` function **combines** two objects into a new object.
   > The `array_unique(array, array, ...)` function **merges** two arrays into one array with unique elements.

Please let me know if you think this could be improved on the CrateDB documentation.

[^1]: https://crate.io/resources/videos/json-object-data-in-cratedb
[^2]: https://crate.io/docs/crate/reference/en/latest/general/builtins/scalar-functions.html#array-functions
[^3]: https://crate.io/docs/crate/reference/en/latest/general/builtins/scalar-functions.html#object-functions

Reply:
Ah. `concat()` does not completely do what I am aiming at here.
```
cr> select time, concat(tags, fields) from mqttkit_2_itest.foo_bar_sensors;
+---------------+------------------------------------------+
|          time | concat(tags, fields)                     |
+---------------+------------------------------------------+
| 1687469154383 | {"humidity": 83.1, "temperature": 42.84} |
+---------------+------------------------------------------+
```

It is nice that it will merge two objects, but now, I would like to destructure the top-level attributes of that single object into individual fields again.